### PR TITLE
[SYCLomatic] Support migration of cub::DiscardOutputIterator

### PR DESCRIPTION
**Pull request by yihwang-nv**
Signed-off-by: Wang, Yihan [yihan.wang@intel.com](mailto:yihan.wang@intel.com)

### CODE REVIEW

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 1]
```diff
@@ -45,6 +45,8 @@ class CheckTemplateArgCount {
   }
 };
 
+// This useful to print a templated type, if explicit template argument number
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: Print a templated type. Pass a STR("") as a template argument for types with no template argument e.g. MyType<>

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 2]
```diff
@@ -57,6 +59,16 @@ std::shared_ptr<TypeLocRewriterFactoryBase> createTypeLocRewriterFactory(
       std::forward<std::function<TemplateArgsT(const TypeLoc)>>(TAsCreator)...);
 }
 
+// This useful to print a no templat type
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: Print a type with no template.

File: clang/lib/DPCT/CallExprRewriter.h [hunk 3]
```diff
@@ -798,6 +798,18 @@ template <class NameT, class... TemplateArgsT> class TemplatedNamePrinter {
   }
 };
 
+/// Print a `type`. This is useful when print no template type.
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/CallExprRewriter.h]: Print a type with no template.

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 4]
```diff
@@ -42,6 +42,7 @@ class TypePrinterRewriter : Printer, public TypeLocRewriter {
   }
 };
 
+/// Print a `type`. This is useful when print template type.
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.h]: Print a templated type.

File: clang/lib/DPCT/TypeLocRewriters.h [hunk 5]
```diff
@@ -54,6 +55,17 @@ class TemplateTypeLocRewriter
             TL, TypeNameCreator(TL), TAsCreator(TL)...) {}
 };
 
+/// Print a `type`. This is useful when print no template type.
```
Comment (AndyCHHuang) [inline clang/lib/DPCT/TypeLocRewriters.h]: Print a type with no template.

**Review by AndyCHHuang — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/TypeLocRewriters.h (hunk 4 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/TypeLocRewriters.h]: done

File: clang/lib/DPCT/CallExprRewriter.h (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/CallExprRewriter.h]: done

Comment (yihanwg) [reply clang/lib/DPCT/CallExprRewriter.h]: done

File: clang/lib/DPCT/TypeLocRewriters.cpp (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: done

File: clang/lib/DPCT/TypeLocRewriters.cpp (hunk 1 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: done

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]
LGTM

**Pull request merged**